design/apistyle-no2.md

```diff
@@ -0,0 +1,496 @@
+const PATH_METADATA_NAME = 'metadata.name';
+const PATH_METADATA_LABELS = 'metadata.labels';
+const PATH_METADATA_ANNOTATIONS = 'metadata.annotations'
+const PATH_EXTERNAL_NAME = 'metadata.annotations[crossplane.io/external-name]';
+const PATH_CONN_SECRET_NAME = 'spec.writeConnectionSecretToRef.name';
+const PATH_CONN_SECRET_NAMESPACE = 'spec.writeConnectionSecretToRef.namespace';
+
+enum NodeSize {
+  SMALL = 'small',
+  MEDIUM = 'medium',
+  LARGE = 'large',
+}
+
+export class ClusterChart extends Chart {
+
+  private id: StringProp = stringProp({
+    path: 'spec.id',
+    description: 'Cluster ID that other objects will use to refer to this cluster',
+    requireAllInPath: true,
+  });
+
+  private networkRef: StringProp = stringProp({
+    path: 'spec.parameters.networkRef.id',
+    description: 'A reference to the Network object that this Cluster should be connected to.',
+    requireAllInPath: true,
+  });
+
+  private nodeCount: IntegerProp = integerProp({
+    path: 'spec.parameters.nodes.count',
+    description: 'Desired node count, from 1 to 100.',
+    requireAllInPath: true,
+  });
+
+  private nodeSize: EnumProp = enumProp({
+    path: 'spec.parameters.nodes.size',
+    description: 'Size of node.',
+    requireAllInPath: true,
+    enum: [
+      NodeSize.SMALL,
+      NodeSize.MEDIUM,
+      NodeSize.LARGE,
+    ]
+  });
+
+  private promVersion: StringProp = stringProp({
+    path: 'spec.parameters.services.operators.prometheus.version',
+    description: 'The version of kube-prometheus-stack chart to install.',
+    requireAllInPath: true,
+  });
+
+  private connectionSecret: ConnectionSecretProps = {
+    defaultNamespace: 'crossplane-system',
+    keys: [`kubeconfig`]
+  };
+
+  constructor(scope: Construct, id: string) {
+    super(scope, id);
+
+    // Cluster XRD
+    const xrdCluster = new crossplane.CompositeResourceDefinition(this, 'cluster-xrd', {
+      name: 'compositeclusters.aws.platform.acme.io',
+      group: 'aws.platform.acme.io',
+      claimKind: {
+        name: 'Cluster',
+        plural: 'clusters',
+      },
+      kind: {
+        name: 'CompositeCluster',
+        plural: 'compositeclusters'
+      },
+      connectionSecret: this.connectionSecret,
+      versions: [
+        {
+          name: 'v1alpha1',
+          served: true,
+          referenceable: true,
+          schemaSpec: [
+            this.id,
+            this.networkRef,
+            this.nodeCount,
+            this.nodeSize,
+            this.promVersion,
+          ],
+        }
+      ],
+    });
+
+    const xrdEKS = this.eks();
+    const xrdServices = this.services();
+
+    // Cluster Composition
+    const clusterComp = new crossplane.Composition(this, 'cluster-comp', xrdCluster, {
+      name: 'compositeclusters.aws.platform.acme.io',
+      metadata: {
+        labels: {
+          provider: 'aws',
+        },
+      },
+    });
+
+    // auto mapped patches for reused props
+    clusterComp.addXrdResource(xrdEKS)
```

Comment:
What's going on here? Not sure I understand

Reply:
the `Composition` has two methods to add resources to a composition:
1) `addResource()` adds an L1 resource from imported Crossplane provider CRDs (e.g. `IamRole`, `NodeGroup`, etc.)
2) 'addXrdResource()' which composes another XRD for a wrapper composition (Acme `Cluster`) whose composition is composed of two other XRDs (`EKS` and `Services`), to avoid a huge monolithic composition and to facilitate reuse.

Since subsets of the API schema in the wrapper `Cluster` XRD needs to be passed down to composed XRDs (often with the same prop path or pulled up a level) there's a lot of boilerplate copying of props from the top-level wrapper XRD down into the composed XRDs. 

Since the props are defined once, and then used in multiple XRDs (sometimes with path munging), `addXrdResource()` suppport auto mapping these fields to reduce the boiler plate code.

For example: this prop def is defined once:
```
private nodeCount: IntegerProp = integerProp({
    path: 'spec.parameters.nodes.count',
    description: 'Desired node count, from 1 to 100.',
    requireAllInPath: true,
  });
```

and then used in both the:
```
// Cluster XRD
    const xrdCluster = new crossplane.CompositeResourceDefinition(this, 'cluster-xrd', {
      name: 'compositeclusters.aws.platform.acme.io',
      group: 'aws.platform.acme.io',
      claimKind: {
        name: 'Cluster',
        plural: 'clusters',
      },
      kind: {
        name: 'CompositeCluster',
        plural: 'compositeclusters'
      },
      connectionSecret: this.connectionSecret,
      versions: [
        {
          name: 'v1alpha1',
          served: true,
          referenceable: true,
          schemaSpec: [
            this.id,
            this.networkRef,
            this.nodeCount,
            this.nodeSize,
            this.promVersion,
          ],
        }
      ],
    });
```

and also in the 
```
// EKS XRD
    const xrd = new crossplane.CompositeResourceDefinition(this, 'eks-xrd', {
      name: 'eks.aws.platform.acme.io',
      group: 'aws.platform.acme.io',
      kind: {
        name: 'EKS',
        plural: 'eks'
      },
      connectionSecret: this.connectionSecret, 
      versions: [
        {
          name: 'v1alpha1',
          served: true,
          referenceable: true,
          schemaSpec: [
            this.id,
            this.networkRef,
            this.nodeCount,
            this.nodeSize,
          ],
        }
      ],
    });
```

when created the prop defs would get a prop id, and then subsequent munging would retain the prop id to all auto mapping any reused prop defs when they child XRD was added to a composition using:
```
// auto mapped patches for reused props
    clusterComp.addXrdResource(xrdEKS)
    .patch(this.id.path, PATH_EXTERNAL_NAME)
    .patchXFormStringFormat(PATH_METADATA_UID, '%s-eks', PATH_CONN_SECRET_NAME)
    .connectionDetailsFromXrd();
```

In this case `addXrdResource(xrdEKS)` auto maps the following for you:
```
clusterComposition.patch('spec.id', 'spec.id');
clusterComposition.patch('spec.parameters.networkRef.id', 'spec.parameters.networkRef.id');
clusterComposition.patch('spec.parameters.nodes.count', 'spec.parameters.nodes.count');
clusterComposition.patch('spec.parameters.nodes.size', 'spec.parameters.nodes.size');
```